### ISSUE TITLE
Postcondition verification fails when returning result of method call

### ISSUE THREAD

**Issue by msridhar** [typetools/checker-framework#2729]
Test case:
```java
import org.checkerframework.checker.nullness.qual.*;
class NullnessTest {
    @Nullable Object f;
    @EnsuresNonNull("this.f")
    public Object retMethodCall() {
        f = new Object();
        //return new Object();
        return createObject();
    }
    static Object createObject() { return new Object(); }
}
```
Compiling with the nullness checker yields an error:
```
NullnessTest.java:5: error: [contracts.postcondition.not.satisfied] the postcondition EnsuresNonNull about 'this.f' of this method is not satisfied
    public Object retMethodCall() {
                  ^
1 error
```
If I comment out `return createObject()` and uncomment `return new Object()` then no error is reported.  This affect other type systems leveraging `@PostconditionAnnotation` qualifiers as well.

**Comment by wmdietl** [comment]
Method `createObject` can have arbitrary side-effects, in particular it can set field `f` to `null`.
So you need to convince the checker that the method isn't doing that, e.g. by marking the method as `@Pure` or adding postconditions to it.
The following type-checks without a warning:

```java
import org.checkerframework.dataflow.qual.Pure;
import org.checkerframework.checker.nullness.qual.*;
class NullnessTest {
    @Nullable Object f;
    @EnsuresNonNull("this.f")
    public Object retMethodCall() {
        f = new Object();
        //return new Object();
        return createObject();
    }
    @Pure
    static Object createObject() { return new Object(); }
}
```

This isn't specific to postconditions, either.
This simple program gives a `dereference.of.nullable` error:

```java
import org.checkerframework.dataflow.qual.Pure;
import org.checkerframework.checker.nullness.qual.*;
class NullnessTest {
    @Nullable Object f;
    void foo() {
        f = new Object();
        createObject();
        f.toString();
    }
    Object createObject() { return new Object(); }
}
```

which goes away when marking `createObject` as pure.

So I think everything is working as intended. Am I missing something?

**Comment by msridhar** [comment]
Yes, you're right.  I realized after a few minutes this is an issue with side effects.  Sorry for the false alarm!
